_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,375 @@
+---
+title: "Clean Unit Tests with Mockito"
+categories: [java]
+date: 2021-03-20 00:00:00 +0100 
+modified: 2021-03-20 00:00:00 +0100 
+author: default 
+excerpt: "Clean Unit Tests with Mockito"
+image:
+  auto: 0098-profile
+---
+
+{% include github-project.html url="https://github.com/silenum/mockito-examples" %}
+
+Mockito allows you to write meaningful and tailored unit tests in Java. It is widely in use, since it is easy to use and
+to create a suitable fixture for your unit test! In this article we will face how to use Mockito, what are Dos and
+Don'ts and learn how to improve your unit test cases.
+
+## Introduction to Mocks
+
+Before we start talking about *Mocks*, we need to understand what *Test Doubles* are. Test Doubles are like stuntmen for
+your test cases. They make the hard work for you while you are shining, doing the nice work. So mocking is used, when we
+want to replace real objects with mocks, whose behaviour is under full control, so that we can advise them what to do.
+
+Here comes [Mockito](https://www.mockito.org) into play. Mockito is a very popular library that supports such testing
+scenarios. Spring also uses Mockito and has own adaptors that provide easy handling for using developers. Don't worry -
+you must not have Spring to use Mockito! Mockito also comes with a standalone library that can be referenced within your
+build tool such as Maven and Gradle.
+
+Consider reading the article [Why Mock?](https://reflectoring.io/spring-boot-mock/#why-mock) for additional information
+about Mocking.
+
+## Test Doubles and Their Duty
+
+In the world of code, there are lots of words for Test Doubles and definitions for their duty. We recommend you to
+define a common language, if different understanding within your team persist.
+
+Here is a little summary of the different types for Test Doubles and how we use them:
+
+| Type  | Description                                                  |
+| ----- | ------------------------------------------------------------ |
+| Dummy | A dummy is an object, that is often used as a parameter for the method under test but without an actual need of it during the test itself. |
+| Fake  | A fake is an object, that has an implementation, but which is restricted and simpler as the real implementation. Due to this restriction, the object is not meant to use in production. |
+| Stub  | A stub is an object, that always returns the same value, regardless of which parameters you provide on a stub's methods. |
+| Mock  | A mock is an object, whose behaviour previously was declared, before the test is run. (This is exactly what Mockito is made for!) |
+| Spy   | A spy is an object, that protocols calls and arguments. It can be queried to create assertions in order to verify the behaviour of the system under test. (Spies are supported by Mockito!) |
+
+Per default Mockito creates mocks of the type *Spy* !
```

Comment:
This will be confusing for readers that don't know Mockito because Mockito differentiates between `Spy` and `Mock`. `Mock`s have some Spy-like attributes (like being able to `verify` method calls), but Mockito `Spy`s spy on real objects.

Perhaps add a short section "Creating Mocks and Spies with Mockito"  after this section here, in which you explain the difference between the two.

Reply:
I removed the confusing sentence and will append a section in "How to Create Mocks with Mockito"